### PR TITLE
Change ChannelsLiveServerTestCase to use a thread instead of a process.

### PR DESCRIPTION
**Pull request by shjohnson-pi**
Part of fix for django/channels#962.

Change ChannelsLiveServerTestCase to use a thread instead of a process.
The thread uses a new Twisted Reactor with a thread local asyncio event loop.
In addition, since the tests are threaded, mocking will work inside tests as well.

### CODE REVIEW

**Comment by shjohnson-pi** [comment]
I just fixed some pyflakes and black warnings.

File: tests/test_example/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+selenium==3.141.0
+service-identity==18.1.0
+six==1.13.0
+sqlparse==0.3.0
+Twisted==19.7.0
+txaio==18.8.1
+urllib3==1.25.6
+wcwidth==0.1.7
+zipp==0.6.0
+zope.interface==4.6.0
```
Comment (jonathan-s) [inline tests/test_example/requirements.txt]: I'm guessing that you used `pip freeze` here. My question here are all these dependencies necessary?

Comment (shjohnson-pi) [reply tests/test_example/requirements.txt]: That was made from a blank virtual environment, then only pip installing what was required to run the example. Then I did a pip freeze to grab all dependencies. This way if two users are comparing example output the only difference should be machine architecture.

Top dependencies are:
- channels-redis
- Django 2.2.7 (I think it fails on Django 3)
- Selenium
- Default dependencies for Daphne and Channels [tests] package.

I added the example for carltongibson, and wasn't sure if he wanted it as a .tar.gz or committed in the repo like this. My Daphne pull request https://github.com/django/daphne/pull/284 goes in more detail about this example.

I hope this answers your question.

**Comment by ckcollab** [comment]
Having mocks work "automagically" would be amazing, any chance this can get merged?

**Comment by shjohnson-pi** [comment]
> Having mocks work "automagically" would be amazing, any chance this can get merged?

@ckcollab This pull request has been put on hold, because of the related pull request at https://github.com/django/daphne/pull/284

It causes daphne tests to be about 3x slower, which @carltongibson thought was unacceptable.

So, although the pull requests cause everything to work correctly on all operating systems, it comes at a cost of speed.

**Comment by carltongibson** [comment]
OK, finally swinging back to this. 

I've merged #1906 to enforce `fork` here, which at least lets test cases work on macOS. (We'll need a fuller solution long-term, and for Windows support.) 

Note pre-4.0 this requires the development version of `daphne` as well. See #1898 

Then looking at a thread-based approach, as per here (which is what Django's LiveServerTestCase uses) as well as a subprocess (rather than multiprocessing) approach is next up to resolve finally. 

The tracking issue is #1207.

**Comment by carltongibson** [comment]
Resolved by https://github.com/django/channels/pull/1922 & v4.0

**Comment by ckcollab** [comment]
Thank you so much :)

**Pull request closed without merging**